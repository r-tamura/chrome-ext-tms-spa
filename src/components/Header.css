:root {
  --header-height: 50px;
}

.company-logo {
  display: block;
  width: 40px;
}


.page-header {
  display: flex;
  align-items: center;
  flex-direction: row;
  padding: 0 5px;
  background-color: var(--primary-color-reverse);
  /* color: var(--white); */
  font-size: var(--font-size-lg);
  font-weight: bolder;
  line-height: var(--header-height);
  
  @apply --z-index-high;
  @apply --box-shadow;

  & a:focus, & a:hover {
    opacity: 0.8
  }

  & .nav {
    height: 80%;
    margin: 5px 0;
  }

  & .nav-link {
    height: 100%;

    &.active {
      background-color: $nav-selected-bg-color;
    }
  }

  @media (--md) {
    /* ロゴ非表示 */
    .nav, .nav-link {
        display: none;
    }

    .menu-btn {
        display: block;
    }
  }
}

/* ヘッダーメニューボタン */
.menu-btn {
    cursor: pointer;
    padding: $header-height/4 10px;
    display: none;
    &:hover {
      background-color: darken($COLOR_HEADER_BG, 10%);
      border-radius: 50%;
    }

    img {
      height: 50%;
    }
}

.header-user {
  /* @extend .clickable; */
  display: flex;
  align-items: center;
  margin-left: auto;

  .person-icon {
    background-image: url(chrome-extension://odbcdehjjdjjnpenjbblpmkkibggblpe/chrome/assets/images/person.svg);
    height: 50%;
    width: 25px;
    display: block;
  }

  &:hover {
    text-decoration: underline;
  }
}

.header-user, .nav-link {
  img {
    display: inline-block;
    height: $header-height;
    width: $header-height;
    padding: $header-height * 0.10 10px;
  }
}

.header-user .dropdown-menu {
  right: 0px;
}
