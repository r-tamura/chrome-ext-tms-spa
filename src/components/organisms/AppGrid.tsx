import styled from "styled-components";

export const AppGrid = styled.div`
  display: grid;
  /* header content (contentの高さは view height - headerの高さ*/
  grid-template-rows: 50px minmax(calc(100vh - 50px), auto);
  /* grid-template-rows: 50px 1fr; */
  grid-template-columns: 200px 1fr;

  & > .app-header {
    grid-column: 1 / 2 span;
  }

  & > .app-content-full {
    grid-column: 1 / 2 span;
  }

  & > .app-navbar {
    grid-column: 1 / 1 span;
  }

  & > .app-content-with-navbar {
    grid-column: 2 / 1 span;
  }
`;
