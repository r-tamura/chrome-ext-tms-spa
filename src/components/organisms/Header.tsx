import React, { FC, SyntheticEvent, useState } from "react";
import Menu, { MenuItem } from "~/components/Menu";
import styled from "styled-components";
import { lighten } from "polished";
import { HIGH } from "~/styles/zIndex";
import { STANDARD } from "~/styles/boxShadow";
import { XL } from "~/styles/font";
import { Theme } from "~/styles/theme";
import { useUser } from "~/stores/hooks";

interface IHeaderUserProps {
  onLogout: () => void;
}

const HeaderUser: FC<IHeaderUserProps> = ({ onLogout, children }) => {
  const [anchorElm, setAnchorElm] = useState(null);

  const menuOpened = anchorElm != null;
  return (
    <UserMenu>
      <ClickableSpan id={"header-user-button"} onClick={handleClick}>
        {children}
      </ClickableSpan>
      <Menu title={"User Menu"} onClose={handleClose} open={menuOpened}>
        <MenuItem id={"logout"} button onClick={handleLogout}>
          Logout
        </MenuItem>
      </Menu>
    </UserMenu>
  );

  function handleClick(e: SyntheticEvent<Element>) {
    setAnchorElm(e.currentTarget);
  }

  function handleClose(_: MouseEvent) {
    setAnchorElm(null);
  }

  function handleLogout(e: SyntheticEvent<Element, MouseEvent>) {
    handleClose(e.nativeEvent);
    onLogout();
  }
};

/**
 * アプリケーションヘッダーコンポーネント
 */
export const AppHeader: FC = () => {
  const { logout, name: username } = useUser();
  return (
    <Root className="app-header">
      <CompanyLogoAnchor href="http://www.telema.jp/">
        <img src="https://www.telema.co.jp/images/logo_sp.svg" alt="logo" />
      </CompanyLogoAnchor>
      <HeaderUser onLogout={logout}>{username}</HeaderUser>
    </Root>
  );
};

const ClickableSpan = styled.span`
  &:hover {
    cursor: pointer;
    text-decoration: none;
  }
`;

const CompanyLogoAnchor = styled.a`
  display: block;
  width: 40px;
  &:focus,
  &:hover {
    opacity: 0.8;
    text-decoration: none;
  }
`;

const Root = styled.header`
  display: flex;
  align-items: center;
  flex-direction: row;
  padding: 0 5px;
  background-color: ${({ theme }) => theme.primaryReverse};
  font-size: ${XL};
  font-weight: bolder;
  line-height: 50px;

  z-index: ${HIGH};
  box-shadow: ${STANDARD};
`;

const UserMenu = styled.div`
  display: flex;
  align-items: center;
  margin-left: auto;
  color: ${({ theme }) => theme.textMain};

  .person-icon {
    background-image: url(chrome-extension://odbcdehjjdjjnpenjbblpmkkibggblpe/chrome/assets/images/person.svg);
    height: 50%;
    width: 25px;
    display: block;
  }

  &:hover {
    color: ${({ theme }: { theme: Theme }) => lighten(0.3, theme.textMain)};
  }
`;
