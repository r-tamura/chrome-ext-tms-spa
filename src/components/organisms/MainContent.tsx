import styled from "styled-components";

export const MainContent = styled.div`
  padding: 10px;
`;
