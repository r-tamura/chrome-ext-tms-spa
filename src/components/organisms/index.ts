export * from "./AppGrid";
export * from "./MainContent";
export * from "./Header";
export * from "./TransExpenseHistory";
export * from "./TransExpenseTemplates";
export * from "./Nav";
