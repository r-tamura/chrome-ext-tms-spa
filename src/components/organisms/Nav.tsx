import React from "react";
import { NavItem } from "~/components/molecules";
import styled from "styled-components";
import { MID } from "~/styles/zIndex";
import { STANDARD } from "~/styles/boxShadow";
import { ThemeProps } from "~/styles/theme";

interface INavProps extends React.Props<{}> {
  path: string;
}

export const Nav: React.SFC<INavProps> = ({ path }) => {
  const pathname = path.replace(/^\/([^/]+).*$/, "$1");
  return (
    <NavStyled id="side-nav-bar">
      <ul>
        {/* リンク一覧 */}
        <li>
          <NavItem
            to={"transportation"}
            selected={"transportation" === pathname}
          >
            交通費
          </NavItem>
        </li>
        <li>
          <NavItem to={"attendance"} selected={"attendance" === pathname}>
            勤怠管理
          </NavItem>
        </li>
        {/* TODO: 資源管理は未実装 */}
        <li>
          <NavItem
            to={"resource"}
            disabled={true}
            selected={"resource" === pathname}
          >
            資源管理
          </NavItem>
        </li>
      </ul>
    </NavStyled>
  );
};

const NavStyled = styled.nav<ThemeProps>`
  background-color: ${({ theme }) => theme.primary};
  z-index: ${MID};
  box-shadow: ${STANDARD};
`;
