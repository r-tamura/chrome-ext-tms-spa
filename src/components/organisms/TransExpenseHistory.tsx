import React, { useState } from "react";
import { find, propEq } from "ramda";
import {
  Button,
  ButtonGroup,
  Panel,
  RequiredInput,
  Modal
} from "~/components/atoms";
import { AddIcon } from "~/components/Icon";
import { CURRENCY } from "~/helpers/_const";
import {
  Project,
  Usage,
  Objective,
  TransExpenseView,
  TransExpenseTemplateView,
  ExpenseId,
  TransExpenseCreateRequest,
  TransExpenseUpdateRequest,
  TransExpense
} from "~/types";

import { Table, Tr, Td } from "~/components/atoms/Table";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faEdit, faTrash } from "@fortawesome/free-solid-svg-icons";
import { useModal } from "~/components/hooks";
import useForm from "react-hook-form";
import { SelectBox } from "../SelectBox";

interface IProps {
  expenses: TransExpenseView[];
  templates: TransExpenseTemplateView[];
  projects: Project[];
  usages: Usage[];
  objectives: Objective[];
  deleteExpense: (expenseId: ExpenseId) => void;
  createExpense: (expense: TransExpenseCreateRequest) => void;
  updateExpense: (expense: TransExpenseUpdateRequest) => void;
  createExpenseFromTemplate: (templateId: string, strdate: string) => void;
}

interface IState {
  selected?: ExpenseId; // 選択中の交通費ID
  useTemplate?: boolean;
}

/**
 * 登録済み交通費リストセクション
 */
export const TransExpenseHistory: React.FC<IProps> = props => {
  const { isOpen: isModalOpen, open, close } = useModal(false);
  const [state, setState] = useState(getInitState(props));

  const { expenses, deleteExpense } = props;
  const { useTemplate } = state;
  return (
    <div>
      {expenses.length > 0 ? <Expenses /> : <p>交通費が登録されていません。</p>}
      <ButtonGroup>
        <Button
          variant={"contained"}
          color={"secondary"}
          onClick={openCreateFromTemplateModal}
        >
          テンプレートから新規登録
        </Button>
        <Button
          id={"create-trans-expense"}
          variant={"circle"}
          color={"primary"}
          onClick={openCreateModal}
        >
          <AddIcon />
        </Button>
      </ButtonGroup>

      {/* 編集モーダル */}
      <EditorModal />
    </div>
  );

  function Expenses() {
    const header = [
      "#",
      "利用日",
      "プロジェクト",
      "利用区分",
      "目的",
      "金額",
      "",
      ""
    ];
    return (
      <div>
        <Table header={header}>
          {expenses.map((e, i) => (
            <Tr key={(() => e.expenseId)()} hover>
              <Td className="record-id">{i + 1}</Td>
              <Td>{e.strdate}</Td>
              <Td>{e.project.name}</Td>
              <Td>{e.usage.name}</Td>
              <Td>{e.objective.name}</Td>
              <Td>
                <strong>{CURRENCY.format(e.cost)}</strong>
              </Td>
              <Td>
                <Button
                  variant="icon"
                  color="primary"
                  onClick={() => openEditModal(e.expenseId)}
                  title="編集"
                >
                  <FontAwesomeIcon icon={faEdit} />
                </Button>
              </Td>
              <Td>
                <Button
                  variant="icon"
                  color="danger"
                  onClick={() => deleteExpense(e.expenseId.toString())}
                  title="削除"
                >
                  <FontAwesomeIcon icon={faTrash} />
                </Button>
              </Td>
            </Tr>
          ))}
        </Table>
      </div>
    );
  }

  /*
   * 交通費新規作成/編集モーダルをレンダーします
   */
  function EditorModal() {
    type FormData = Record<keyof TransExpense, string>;
    const {
      register,
      handleSubmit,
      clearError,
      errors,
      setError,
      formState
    } = useForm<FormData>();
    const {
      expenses,
      projects,
      usages,
      objectives,
      createExpense,
      updateExpense
    } = props;
    const { selected } = state;

    function handleCancelClick(e: React.MouseEvent) {
      e.preventDefault();
      e.stopPropagation();
      closeModal();
    }

    function onSubmit({
      strdate,
      projectId,
      usageId,
      objectiveId,
      customer,
      from,
      to,
      cost: costStr
    }: FormData) {
      let cost;
      cost = Number.parseInt(costStr, 10);
      if (Number.isNaN(cost)) {
        setError(
          "cost",
          "NaN",
          "有効な金額ではありません、正の整数値を入力してください"
        );
        return;
      }
      if (cost < 0) {
        setError("cost", "NaN", "正の整数値を入力してください");
        return;
      }

      if (selected) {
        updateExpense({
          expenseId,
          strdate,
          projectId,
          usageId,
          objectiveId,
          customer,
          from,
          to,
          cost
        });
      } else {
        createExpense({
          strdate,
          projectId,
          usageId,
          objectiveId,
          customer,
          from,
          to,
          cost
        });
      }
      closeModal();
    }

    const defaultExpense: TransExpenseView = {
      expenseId: null,
      strdate: "",
      project: projects[0],
      usage: usages[0],
      objective: objectives[0],
      customer: "",
      from: "",
      to: "",
      cost: 0
    };

    const {
      expenseId,
      strdate,
      project,
      usage,
      objective,
      customer,
      from,
      to,
      cost
    }: TransExpenseView =
      find(propEq("expenseId", selected), expenses) || defaultExpense;
    const costStr = cost.toString();
    return (
      <Modal isOpen={isModalOpen} onRequestClose={() => close()}>
        <Panel>
          <form onSubmit={handleSubmit(onSubmit)}>
            <RequiredInput
              name={"strdate"}
              label={"日付"}
              placeholder={"ex: 20190401"}
              error={errors.strdate}
              defaultValue={strdate}
              register={register}
              clearError={clearError}
            />
            <SelectBox
              name="projectId"
              label="プロジェクトコード *"
              defaultValue={project.projectId}
              options={{
                items: projects,
                value: (project: Project) => project.projectId,
                label: (project: Project) => project.name
              }}
              ref={register({ required: true })}
            />
            <SelectBox
              name="usageId"
              label="利用区分 *"
              defaultValue={usage.usageId}
              options={{
                items: usages,
                value: (usage: Usage) => usage.usageId,
                label: (usage: Usage) => usage.name
              }}
              ref={register({ required: true })}
            />
            <SelectBox
              name="objectiveId"
              label="目的コード *"
              defaultValue={objective.objectiveId}
              options={{
                items: objectives,
                value: (objective: Objective) => objective.objectiveId,
                label: (objective: Objective) => objective.name
              }}
              ref={register({ required: true })}
            />
            <RequiredInput
              name={"customer"}
              label={"顧客先"}
              placeholder={"ex: XXXX 株式会社"}
              error={errors.customer}
              defaultValue={customer}
              register={register}
              clearError={clearError}
            />
            <RequiredInput
              name={"from"}
              label={"乗車地"}
              placeholder={"ex: JR 千葉駅"}
              error={errors.from}
              defaultValue={from}
              register={register}
              clearError={clearError}
            />
            <RequiredInput
              name={"to"}
              label={"降車地"}
              placeholder={"ex: JR 東京駅"}
              error={errors.to}
              defaultValue={to}
              register={register}
              clearError={clearError}
            />
            <RequiredInput
              name={"cost"}
              label={"金額"}
              placeholder={"ex: 500"}
              error={errors.cost}
              defaultValue={costStr}
              register={register}
              clearError={clearError}
            />
            <ButtonGroup>
              <Button variant="contained" onClick={handleCancelClick}>
                キャンセル
              </Button>
              <Button
                variant="contained"
                color="primary"
                disabled={!formState.isValid}
              >
                {expenseId ? "更新" : "作成"}
              </Button>
            </ButtonGroup>
          </form>
        </Panel>
      </Modal>
    );
  }

  function renderCreateFromTemplateModal() {
    const { templates, createExpenseFromTemplate } = props;
    return (
      // <Modal
      //   isOpen={isModalOpen}
      //   formItems={[
      //     {
      //       type: FormItemType.TEXT,
      //       name: "strdate",
      //       label: "日付",
      //       value: ""
      //     },
      //     {
      //       type: FormItemType.SELECT,
      //       name: "templateId",
      //       label: "テンプレート名",
      //       value: "",
      //       options: {
      //         items: templates,
      //         valueKey: "templateId",
      //         labelKey: "templateName"
      //       }
      //     }
      //   ]}
      //   onClose={closeModal}
      //   onOKClick={({
      //     templateId,
      //     strdate
      //   }: {
      //     templateId: string;
      //     strdate: string;
      //   }) => createExpenseFromTemplate(templateId, strdate)}
      // />
      <div> history template</div>
    );
  }

  function getInitState(props: IProps): IState {
    return { useTemplate: false, selected: undefined };
  }

  /* モーダルクローズ */
  function closeModal() {
    setState({
      useTemplate: false,
      selected: null
    });
    close();
  }

  /* 編集モーダル */
  function openEditModal(expenseId: ExpenseId) {
    setState({
      useTemplate: false,
      selected: expenseId
    });
    open();
  }

  /* 新規登録モーダル */
  function openCreateModal() {
    setState({
      useTemplate: false,
      selected: null
    });
    open();
  }

  /* テンプレートから新規登録モーダル */
  function openCreateFromTemplateModal() {
    setState({
      useTemplate: true,
      selected: null
    });
    open();
  }
};
