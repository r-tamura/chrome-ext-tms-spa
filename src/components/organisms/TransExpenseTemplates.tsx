import React, { useState } from "react";
import useForm from "react-hook-form";
import { find, propEq } from "ramda";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrash, faEdit } from "@fortawesome/free-solid-svg-icons";
import { Modal } from "~/components/atoms/Modal";
import { Button, ButtonGroup, Panel, Input } from "~/components/atoms";
import { AddIcon } from "~/components/Icon";
import { CURRENCY } from "~/helpers/_const";
import {
  Project,
  Usage,
  Objective,
  TransExpenseTemplateView,
  TransExpenseTemplateCreateRequest,
  TransExpenseTemplateUpdateRequest,
  TransExpenseTemplate,
  TemplateId
} from "~/types";
import { Table, Tr, Td } from "~/components/atoms/Table";
import { useModal } from "~/components/hooks";
import { SelectBox } from "../SelectBox";
import { ReactHookFormError } from "react-hook-form/dist/types";

interface IProps {
  templates: TransExpenseTemplateView[];
  projects: Project[];
  usages: Usage[];
  objectives: Objective[];
  createExpenseTemplate: (template: TransExpenseTemplateCreateRequest) => void;
  deleteExpenseTemplate: (templateId: string) => void;
  updateExpenseTemplate: (template: TransExpenseTemplateUpdateRequest) => void;
}

/**
 * 交通費テンプレートコンポーネント
 */
export const TransExpenseTemplates: React.FC<IProps> = ({
  templates,
  projects,
  usages,
  objectives,
  createExpenseTemplate,
  updateExpenseTemplate,
  deleteExpenseTemplate
}) => {
  const { isOpen: isModalOpen, open, close } = useModal(false);
  const [selected, setSelected] = useState<TemplateId | null>(null);
  return (
    <div>
      {/* テンプレート一覧 */}
      {templates.length > 0 ? (
        <TemplatesTable />
      ) : (
        <p>交通費テンプレートがブラウザに登録されていません。登録しますか？</p>
      )}
      {/* 新規登録ボタン */}
      <ButtonGroup>
        <Button
          id={"create-trans-expense-template"}
          color={"primary"}
          variant={"circle"}
          onClick={openCreateModal}
        >
          <AddIcon />
        </Button>
      </ButtonGroup>

      {/* 編集モーダル */}
      <EditorModal />
    </div>
  );

  function TemplatesTable() {
    const header = [
      "#",
      "テンプレート名",
      "プロジェクト",
      "利用区分",
      "目的",
      "金額",
      "",
      ""
    ];

    function createHandleDelete(templateId: TemplateId) {
      return () => deleteExpenseTemplate(templateId);
    }
    return (
      <div>
        <Table header={header}>
          {templates.map((t, i) => (
            <Tr key={t.templateId} hover>
              <Td className="record-id">{i + 1}</Td>
              <Td>{t.templateName}</Td>
              <Td>{t.project.name}</Td>
              <Td>{t.usage.name}</Td>
              <Td>{t.objective.name}</Td>
              <Td>{CURRENCY.format(t.cost)}</Td>
              <Td>
                <Button
                  onClick={() => openEditModal(t.templateId)}
                  variant="icon"
                  color="primary"
                  title="編集"
                >
                  <FontAwesomeIcon icon={faEdit} />
                </Button>
              </Td>
              <Td>
                <Button
                  variant="icon"
                  color="danger"
                  title="削除"
                  onClick={createHandleDelete(t.templateId)}
                >
                  <FontAwesomeIcon icon={faTrash} />
                </Button>
              </Td>
            </Tr>
          ))}
        </Table>
      </div>
    );
  }

  /*
   * 交通費新規作成/編集モーダルをレンダーします
   */
  function EditorModal() {
    type FormData = Record<keyof TransExpenseTemplate, string>;
    const {
      register,
      handleSubmit,
      errors,
      setError,
      formState,
      clearError
    } = useForm<FormData>();
    function onSubmit({
      templateName,
      projectId,
      usageId,
      objectiveId,
      customer,
      from,
      to,
      cost
    }: FormData) {
      let costNumber;
      try {
        costNumber = Number.parseInt(cost, 10);
        if (costNumber < 0) {
          setError("cost", "NaN", "正の整数値を入力してください");
          return;
        }
      } catch (e) {
        setError("cost", "NaN", "正の整数値を入力してください");
        return;
      }

      if (selected) {
        // 更新
        updateExpenseTemplate({
          templateId: find<TransExpenseTemplateView>(
            propEq("templateId", selected),
            templates
          ).templateId,
          templateName,
          projectId,
          usageId,
          objectiveId,
          customer,
          from,
          to,
          cost: costNumber
        });
      } else {
        // 新規作成
        createExpenseTemplate({
          templateName,
          projectId,
          usageId,
          objectiveId,
          customer,
          from,
          to,
          cost: costNumber
        });
      }
      close();
    }

    function handleCancelClick(event: React.MouseEvent) {
      event.stopPropagation();
      event.preventDefault();
      close();
    }

    const defaultTemplate: TransExpenseTemplateView = {
      templateId: "",
      templateName: "",
      project: null,
      usage: null,
      objective: null,
      customer: "",
      from: "",
      to: "",
      cost: 0
    };
    const {
      templateName,
      project,
      usage,
      objective,
      customer,
      from,
      to,
      cost
    } = selected
      ? find(propEq("templateId", selected), templates)
      : defaultTemplate;
    const costStr = cost === 0 ? "" : cost.toString();

    function RequiredInput({
      label,
      ...otherProps
    }: {
      name: keyof FormData;
      label: string;
      placeholder: string;
      defaultValue: string;
      error: ReactHookFormError;
    }) {
      return (
        <Input
          {...otherProps}
          error={!!otherProps.error}
          label={label + " *"}
          ref={register({ required: `${label}は必須入力です` })}
          onFocus={() => clearError(otherProps.name)}
        />
      );
    }

    return (
      <Modal isOpen={isModalOpen} onRequestClose={handleModalClose}>
        <Panel>
          <form onSubmit={handleSubmit(onSubmit)}>
            <RequiredInput
              name="templateName"
              label="テンプレート名"
              placeholder="ex: 定期代(JR千葉駅)"
              defaultValue={templateName}
              error={errors.templateName}
            />
            <SelectBox
              name="projectId"
              label="プロジェクトコード *"
              defaultValue={project ? project.projectId : ""}
              options={{
                items: projects,
                value: (project: Project) => project.projectId,
                label: (project: Project) => project.name
              }}
              ref={register({ required: true })}
            />
            <SelectBox
              name="usageId"
              label="利用区分 *"
              defaultValue={usage ? usage.usageId : ""}
              options={{
                items: usages,
                value: (usage: Usage) => usage.usageId,
                label: (usage: Usage) => usage.name
              }}
              ref={register({ required: true })}
            />
            <SelectBox
              name="objectiveId"
              label="目的コード *"
              defaultValue={objective ? objective.objectiveId : ""}
              options={{
                items: objectives,
                value: (objective: Objective) => objective.objectiveId,
                label: (objective: Objective) => objective.name
              }}
              ref={register({ required: true })}
            />
            <RequiredInput
              name="customer"
              label="顧客先"
              placeholder="ex: XX株式会社"
              defaultValue={customer}
              error={errors.customer}
            />
            <RequiredInput
              name="from"
              label="乗車地"
              placeholder="ex: JR千葉駅"
              defaultValue={from}
              error={errors.from}
            />
            <RequiredInput
              name="to"
              label="降車地"
              placeholder="ex: JR東京駅"
              defaultValue={to}
              error={errors.to}
            />
            <RequiredInput
              name="cost"
              label="金額"
              placeholder="500"
              defaultValue={costStr}
              error={errors.cost}
            />
            <ButtonGroup>
              <Button variant="contained" onClick={handleCancelClick}>
                キャンセル
              </Button>
              <Button
                variant="contained"
                color="primary"
                disabled={!formState.isValid}
              >
                作成/更新
              </Button>
            </ButtonGroup>
          </form>
        </Panel>
      </Modal>
    );
  }

  function handleModalClose() {
    close();
  }

  /* 編集モーダルオープン */
  function openEditModal(templateId: string) {
    setSelected(templateId);
    open();
  }

  /* 新規登録ボタンクリックイベント */
  function openCreateModal(e: React.MouseEvent<HTMLButtonElement>) {
    open();
    setSelected("");
  }
};
