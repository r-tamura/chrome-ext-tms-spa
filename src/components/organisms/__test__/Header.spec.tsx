import React from "react";
import { mount } from "enzyme";
import { AppHeader } from "~/components/organisms/Header";
import { ThemeProvider } from "styled-components";
import { telema } from "~/styles/theme";
import * as hooks from "~/stores/hooks";
import toJson from "enzyme-to-json";

jest.mock("../../../stores/hooks");
const mockedHooks = hooks as jest.Mocked<typeof hooks>;

describe("<AppHeader />", () => {
  it("shallow render snapshot", () => {
    const mockLogoutUser = jest.fn();
    mockedHooks.useUser.mockReturnValue({
      isAuthenticated: true,
      isFetching: false,
      name: "testuser",
      login: jest.fn(),
      logout: mockLogoutUser
    });
    const $ = mount(
      <ThemeProvider theme={telema}>
        <AppHeader />
      </ThemeProvider>
    );

    expect(toJson($)).toMatchSnapshot();
  });

  it("shallow rendering", () => {
    const mockLogoutUser = jest.fn();
    mockedHooks.useUser.mockReturnValue({
      isAuthenticated: true,
      isFetching: false,
      name: "testuser",
      login: jest.fn(),
      logout: mockLogoutUser
    });
    const wrapper = mount(
      <ThemeProvider theme={telema}>
        <AppHeader />
      </ThemeProvider>
    );

    const $a = wrapper.find("a");
    expect($a.length).toBe(1);
    expect($a.props().href).toBe("http://www.telema.jp/");

    const $HeaderUser = wrapper.find("HeaderUser");
    expect($HeaderUser.prop("onLogout")).toEqual(mockLogoutUser);
    expect($HeaderUser.children().text()).toBe("testuser");
  });

  describe("", () => {
    const mockLogoutUser = jest.fn();
    mockedHooks.useUser.mockReturnValue({
      isAuthenticated: true,
      isFetching: false,
      name: "testuser",
      login: jest.fn(),
      logout: mockLogoutUser
    });
    const $ = mount(
      <ThemeProvider theme={telema}>
        <AppHeader />
      </ThemeProvider>
    );

    it("子要素が2", () => {
      expect($.find("Menu").children().length).toBe(0);
    });

    it("クリックイベント", () => {
      $.find("span#header-user-button")
        .at(0)
        .simulate("click");
      expect($.find("Menu").children().length).toBe(1);
      $.find("button#logout").simulate("click");
      expect(mockLogoutUser.mock.calls.length).toBe(1);
    });
  });
});
