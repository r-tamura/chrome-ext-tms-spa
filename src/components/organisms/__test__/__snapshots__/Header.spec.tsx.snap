// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppHeader /> shallow render snapshot 1`] = `
<ThemeProvider
  theme={
    Object {
      "black": "rgb(5, 5, 5)",
      "buttonDefault": "rgba(0, 0, 0, 0.08)",
      "danger": "#D32F2F",
      "fontFamily": "\\"M PLUS Rounded 1c\\", sans-serif",
      "formBorderColor": "rgba(0, 0, 0, 0.16)",
      "formBorderColorDisabled": "rgba(0, 0, 0, 0.08)",
      "formBorderColorError": "#D32F2F",
      "formBorderColorFocused": "hsl(206, 100%, 31%)",
      "hoveredBg": "rgba(0, 0, 0, 0.03)",
      "inputPlaceHolder": "rgba(0, 0, 0, 0.09)",
      "primary": "hsl(206, 100%, 31%)",
      "primaryDark": "hsl(208, 80%, 27%)",
      "primaryHeavyDark": "hsl(208, 80%, 27%)",
      "primaryHeavyLight": "hsl(199, 91%, 64%);",
      "primaryLight": "hsl(201, 98%, 41%);",
      "primaryReverse": "hsla(0, 0%, 99%, 1);",
      "primaryReverseDark": "hsla(0, 0%, 99%, 0.5);",
      "secondary": "#689F38",
      "textGray": "hsla(0, 5%, 5%, .25)",
      "textMain": "hsl(0, 5%, 5%)",
      "white": "rgb(253, 253, 253)",
    }
  }
>
  <Component>
    <styled.header
      className="app-header"
    >
      <StyledComponent
        className="app-header"
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-kgoBCf",
              "isStatic": false,
              "lastClassName": "cEKvpN",
              "rules": Array [
                "
  display: flex;
  align-items: center;
  flex-direction: row;
  padding: 0 5px;
  background-color: ",
                [Function],
                ";
  font-size: ",
                "21px",
                ";
  font-weight: bolder;
  line-height: 50px;

  z-index: ",
                "999",
                ";
  box-shadow: ",
                "0px 0px 5px #222222",
                ";
",
              ],
            },
            "displayName": "styled.header",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-kgoBCf",
            "target": "header",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <header
          className="app-header sc-kgoBCf cEKvpN"
        >
          <styled.a
            href="http://www.telema.jp/"
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-chPdSV",
                    "isStatic": true,
                    "lastClassName": "iJSWDz",
                    "rules": Array [
                      "
  display: block;
  width: 40px;
  &:focus,
  &:hover {
    opacity: 0.8;
    text-decoration: none;
  }
",
                    ],
                  },
                  "displayName": "styled.a",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-chPdSV",
                  "target": "a",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              href="http://www.telema.jp/"
            >
              <a
                className="sc-chPdSV iJSWDz"
                href="http://www.telema.jp/"
              >
                <img
                  alt="logo"
                  src="https://www.telema.co.jp/images/logo_sp.svg"
                />
              </a>
            </StyledComponent>
          </styled.a>
          <HeaderUser
            onLogout={[MockFunction]}
          >
            <styled.div>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-kGXeez",
                      "isStatic": false,
                      "lastClassName": "ekWWdm",
                      "rules": Array [
                        "
  display: flex;
  align-items: center;
  margin-left: auto;
  color: ",
                        [Function],
                        ";

  .person-icon {
    background-image: url(chrome-extension://odbcdehjjdjjnpenjbblpmkkibggblpe/chrome/assets/images/person.svg);
    height: 50%;
    width: 25px;
    display: block;
  }

  &:hover {
    color: ",
                        [Function],
                        ";
  }
",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-kGXeez",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-kGXeez ekWWdm"
                >
                  <styled.span
                    id="header-user-button"
                    onClick={[Function]}
                  >
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-kAzzGY",
                            "isStatic": true,
                            "lastClassName": "fRLZpl",
                            "rules": Array [
                              "
  &:hover {
    cursor: pointer;
    text-decoration: none;
  }
",
                            ],
                          },
                          "displayName": "styled.span",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-kAzzGY",
                          "target": "span",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      id="header-user-button"
                      onClick={[Function]}
                    >
                      <span
                        className="sc-kAzzGY fRLZpl"
                        id="header-user-button"
                        onClick={[Function]}
                      >
                        testuser
                      </span>
                    </StyledComponent>
                  </styled.span>
                  <Menu
                    onClose={[Function]}
                    open={false}
                    title="User Menu"
                  />
                </div>
              </StyledComponent>
            </styled.div>
          </HeaderUser>
        </header>
      </StyledComponent>
    </styled.header>
  </Component>
</ThemeProvider>
`;
